### DISCUSSION TITLE
Data durability and consistency with datascript and datahike

### DISCUSSION THREAD

**Discussion by arkRedM** [athensresearch/athens#1082] [General]
I will position this close to the problem statement of real-time collaboration than anything else.

Setting up context, the first part of rtc that you are seeing in athens-backend and app is a way to stream transactions. We used
1. datahike as single write
2. datascript as client db
3. ws to stream txns

But the current one won't cut it. More specifically the data flow is not enough to ensure data durability and consistency.

I'll talk about it in the loom
1. Durability - https://www.loom.com/share/2a73776b2c42489b9e4df80aaa94acfe
2. Consistency - https://www.loom.com/share/8613cc912cb44d7482125b1f0caee3a3

PR backend: https://github.com/athensresearch/athens-backend/pull/2
PR athens app: https://github.com/athensresearch/athens/pull/1075

Let me know if I couldn't get the ideas out correctly or if something feels fragmented.

**Comment by jefftangx** [comment]
Just to make I understand the loom videos correctly, the new changes are that the client db must now wait until the server acknowledges the previous transaction before sending new transactions. This is sort of like an `ACK` pattern a la TCP?

**Reply by arkRedM** [reply]
Not just ack(that can be only be done with exact copy of data -- like I talked in loom). here we are making aware of final data state from backend with updated remote-ids.

**Comment by jefftangx** [comment]
This (v2) is more responsive than v1! Did some stress testing with the newest commits on backend (https://github.com/athensresearch/athens-backend/pull/2/commits/260e54477f9eee4e2e6d8dbb13c6b3312acbf141) and frontend (https://github.com/athensresearch/athens/pull/1075/commits/ec83342801ad9cc57c08c411ea36bfa9ecbd600f).

I know frontend and backend latest updates are both drafts, but there are still sync issues. Demo: https://www.loom.com/share/3f5aa016d830472da966a40e4f94dab0

I remember @neotyk said a single transactor was the way to go. Doesn't seem like we are doing that right now. Any reason to not try that for now? Will be slower but consistency seems more important for an MVP.

**Reply by arkRedM** [reply]
> single transactor

backend is still a single transactor. for the client-side we are using the second db to sync and add txns irrespective of curr user view state. Tangling view, data and sync was not great and lead to all weird things. Maybe we can ditch `db-with-remote`. Need to think more.

**Reply by arkRedM** [reply]
> but there are still sync issues

First one(bug-1) was interesting. Not sure how that happened.
Rest I'm actively looking into.

**Reply by jefftangx** [reply]
> backend is still a single transactor. for the client-side we are using the second db to sync and add txns irrespective of curr user view state. Tangling view, data and sync was not great and lead to all weird things. Maybe we can ditch `db-with-remote`. Need to think more.

I think this was the motivation for making the client-side db read-only. Anyway to test that out?

**Comment by jefftangx** [comment]
Setup a remote EC2 instance. Anyone can connect with the following credentials:

address: `ec2-3-16-89-123.us-east-2.compute.amazonaws.com`
token: `x`

A less powerful server (and further networking) seems to make it easier for clients/db to get out of sync.

**Comment by jefftangx** [comment]
Synced up with @neotyk. Notes from our call.

# Problems

- [ ] fix shape of data: tx-id, user-id in different places #1141
- [ ] redo construction of context-dependent tx-data ~> #1140 
  - [ ] define pure verbs for blocks, pages, links, orders, strings, children, e.g. `indent`, `split-block`, `write-string`
  - [ ] centralize all side effects to the very end at the edge as a CLJC library
  - [ ] the CLJC library manages all context
- [ ] with number 2 done, we can separate frontend and backend so local or remote dbs resolution is easy

# Other Notes

  the problem is that each tx expects the current state of datascript to be correct
  this is why we had to debounce enter
  alex: removed debounce and "Locked" the current block

how do we construct tx-data?
much of the time, tx-data looks at surrounding context -> siblings, parents, kids, target, origin
is it possible to "pub" messages that do not look at surrounding context, and then the single writer db determines final datascript state?

is it possible to send context-independent, declarative transactions? then let backend resolve "side effects"

"side effects"
- order (enter, indent)
- block/string (join)
- walk-string
  - determine if a page or reference exists or not
    * "" -> "[[test]]" if test page doesn't exist yet, create page and reference
    * "" -> "[[test]]" if test page exists, create reference
    * "[[test]]" -> "" delete reference
    * "[[test]]" -> "" and it's only the reference, delete reference and delete page
    * "((block ref))" create reference if it doesn't exist yet

**Reply by jefftangx** [reply]
Update from standup today. We already have a language for pure transactions (I call them `verbs` above). These are re-frame events! The current plan this week is to see if we can the tx construction and execution towards the backend.

>  with number 2 done, we can separate frontend and backend so local or remote dbs resolution is easy

We don't have to rewrite the logic to pure and then move to the backend. It already is somewhat pure a la re-frame events, so we can start moving to the backend now. Once we have a single-writer (i.e. frontend datascript is read-only), we can refactor tx's to be more pure, if it's even needed.

**Reply by arkRedM** [reply]
> is it possible to send context-independent, declarative transactions? 

We only send datoms so it should be fine. I am not sure of how what got discussed, but just adding a few thoughts here.

1. locking
Obscenely bad ux. What about “locking” for other users?

2. Special case
No where should backend have any idea about what user did(enter/indent). Backend should only get and modify based on data only. I mean actions are ever increasing and will be uncontrollably long(look at how many we have in app itself)


> is it possible to "pub" messages that do not look at surrounding context, and then the single writer db determines final datascript state?

The pub message is result of applying change on the db, so we don’t really send any contextual information that breaks this abstraction in any way, probably a wrong place to look at.

**Reply by jefftangx** [reply]
> locking
> Obscenely bad ux. What about “locking” for other users?

I think it's not locking per se but rejecting stale transactions.

> Backend should only get and modify based on data only.

This works if all dbs have the same state. If two writes happen at the same time from 2+ clients, it's very easy for the transaction construction (which is context-dependent) to create a race condition.

**Reply by arkRedM** [reply]
> This works if all dbs have the same state.

2 writes cannot happen to same db at the same time. It’s one after other.  So what diff will it make if actions are sent or txns are sent? It will result in same race conditions nonetheless, right?

actions are way too many already and all actions convert to a common format(datoms — non contextual txn data) which is better and easily understood

**Reply by jefftangx** [reply]
> So what diff will it make if actions are sent or txns are sent? It will result in same race conditions nonetheless, right?

It actions are sent, tx-data can be constructed by the final source of truth - server database.

Right now it is user interaction -> reframe event -> tx-data construction -> tx-data execution. Because tx-data construction right now happens on the client, it's very easy, for instance, for block orders to get messed up.

If the events that deal with datascript tx-data construction are passed to the backend and there is a single writer, we can keep data pure until the very edge, keeping tx-data construction more non-conflicting. Thoughts @neotyk @whilo

**Reply by arkRedM** [reply]
> Because tx-data construction right now happens on the client, it's very easy, for instance, for block orders to get messed up.

Doesn’t it make sense to happen here; Where data is created due to user event. Even then sending actions doesn’t ensure durability (in this case - ordering — two clients can issue an action saying “create a first child of X”)

link to first comment:
https://github.com/athensresearch/athens/discussions/1082#discussion-3343075

**Reply by jefftangx** [reply]
> Even then sending actions doesn’t ensure durability (in this case - ordering — two clients can issue an action saying “create a first child of X”)

This is exactly what we want I believe! A single backend writer could reject or resolve this more easily.

This is also the motivation for creating purer, context-independent events. The events are often not literally "create first child" but "enter", "indent", "delete." (There will certainly be edge cases.) Queueing these sorts of events/verbs uses domain language, and therefore seems to encourage intuitive resolutions.

The problem isn't conflicts per se. the problem is datascript db getting into a funky state - multiple blocks with same parent but duplicate order, or a block with multiple parents - which happens pretty easily with tx-data construction happening on client dbs.

The single transaction log ensures reconstruction of non-funky state and. Last writer is fine because undo could just access overwritten events. History is always non-funky. 

This is my understanding of the problem! I hope "funky" makes sense here, can't think of a better word right now :)

**Reply by arkRedM** [reply]
> This is also the motivation for creating purer, context-independent events. 

a set of ctx independent events is exactly same as txn data.

> the problem is datascript db getting into a funky state

funky data/bad state is what I am running into as well. All the bad state is because of invariants(basically our data has some rules for existing, ordering or one parent etc) but that doesn’t encourage need for actions as abstraction (current draft branch has code to deal with data rules/enforcing via invariants)

**Reply by jefftangx** [reply]
> a set of ctx independent events is exactly same as txn data.

This would be true, except tx-data is currently constructed with context, which is what leads to dependence and bad state. 

> All the bad state is because of invariants(basically our data has some rules for existing, ordering or one parent etc) but that doesn’t encourage need for actions as abstraction (current draft branch has code to deal with data rules/enforcing via invariants)

This is actually a very good point. Bad data could still exist, so this doesn't necessarily address the root cause, it just centralizes the tx-data construction and execution and gives that responsibility to the single backend writer.

It would be really cool if we could enforce the structure of the data ("implicit rules") with a schema, but I don't think datascript is capable of this kind of schema enforcement natively. This seems like an app/client-level resolution. 

I haven't read through the rules you've written yet. It would be helpful to hear what your approach is, because it seems viable.

**Reply by arkRedM** [reply]
Current approach is relatively naive.

first we calculate final state as if new txn is applied and calculate all possible invariants for all entities involved. If some entity requires a change(addition or retraction) we add it to the list of transactions to essentially “correct” data before committing. This way data is always in the correct state irrespective of race conditions. 

The current bad states we are seeing are probably because of bugs in code

**Reply by jefftangx** [reply]
If each client can correct tx-data, how do we know they would correct them and eventually arrive at the same state?

**Reply by neotyk** [reply]
- In a situation where:
    - Page has a list of 10 blocks
    - 2 clients are connected
    - both clients in the same time create new second block (by pressing Enter at the end of 1st block)
        - tx from client to server approach:
            - Each client adds new block and reorders 9 blocks + updates parent
            - Each client applies  these transaction to their best known state of DB
            - Both clients send tx log to backend (`:db/id`s are managed)
            - Whichever clients data arrives first get's applied to backend DB
            - Now interesting things happen when second client's tx log arrives
                - We'll create another second block
                - reordering of remaining 9 blocks should be the same
                - parent update from tx log does not include knowledge of changes made by 1st client
                - We'll end up with 1st clients 2nd block orphaned
        - verbs/events to server approach:
            - Both clients send an event to server `:create-new-block-event`
                - This event includes knowledge of last seen transaction by this client
            - Server receives 1st client's event
                - resolves `:create-new-block-event` in context of only true DB and transacts
                - tx-log get's broadcasted to connected clients
            - Server receives 2nd client's event
                - Compares `last-seen-tx` by client to real `last-tx` and rejects the request because out of sync, communicates what is global `last-tx`
            - 2nd client receives rejection with `last-tx`
                - Client compares it's `last-seen-tx`
                    - if already up to date sends same event again 
                        - (because broadcast was received before rejection)
                    - If not up to date, register `tx-wait-event` that when awaited tx arrives get's fired and send the event/verb again

- I don't see a solution to provide consistency in the 1st scenario where we share tx-log from client to server, what did I miss?
- Maybe I'm missing something in my approach.

**Reply by arkRedM** [reply]
First of all two approaches are not different in terms of actions vs txns being sent. Actions or txns the end result is the same(maybe some data format but nothing out of normal). The only change that is happening here is rejecting a txn based on last-tx-id.

In scenario described the most expected solution is both new blocks get added without losing any data. It makes more logical sense as both are independent blocks of a parent. I am not sure locking/rejecting such a txn even makes sense. There is actually no conflict in data layer at all and no implicit reason to block the txn.

 Similarly with same block data change and locking. Last one should always win

**Reply by neotyk** [reply]
If we both modify same block, and your tx got transacted 1st, I would expect my tx to be rejected, and not overwrite your changes.

**Reply by arkRedM** [reply]
Why reject when there is no conflict?

**Comment by arkRedM** [comment]
server: 45.79.158.23:3010
token: x

Running with most minimal setup(need some testing and rtc sync testing here -- server in NJ -- expect delays not bugs). Please record loom/some info that can help debug issues. This will help productionize RTC faster
Thanks.

Note: please check front end from dev https://github.com/athensresearch/athens/pull/1075/files

**Reply by jefftangx** [reply]
@arkRedM could you add your thoughts to the newest frontend and backend PR? I saw you removed datsync, not sure if there are other larger changes being made https://github.com/athensresearch/athens-backend/pull/2/files. Communication very helpful once there are multiple people thinking and working on something.

**Reply by arkRedM** [reply]
Datsync was not adding much value sticking around in the backend. I just picked fundamentals and some code from datsync to manage how txns are handled and removed the dep.

Regarding front-end datsync still manages most of remote id stuff so keeping it there until I find a cleaner abstraction. But it's non-blocking in the current set. 

Other major changes we have been discussing in the backend channel and this thread.
